src/ray/object_manager/pull_manager.cc

```diff
@@ -40,27 +40,30 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
   if (it == pull_requests_.end()) {
     return;
   }
+  auto &request = it->second;
   // Reset the list of clients that are now expected to have the object.
   // NOTE(swang): Since we are overwriting the previous list of clients,
   // we may end up sending a duplicate request to the same client as
   // before.
   it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());
-  if (!spilled_url.empty()) {
-    RAY_LOG(DEBUG) << "OnLocationChange " << spilled_url << " num clients "
-                   << client_ids.size();
-    // Try to restore the spilled object.
-    restore_spilled_object_(object_id, spilled_url,
-                            [this, object_id](const ray::Status &status) {
-                              // Fall back to fetching from another object manager.
-                              if (!status.ok()) {
-                                TryPull(object_id);
-                              }
-                            });
-  } else {
-    // New object locations were found, so begin trying to pull from a
-    // client. This will be called every time a new client location
-    // appears.
-    TryPull(object_id);
+  if (request.inflight_pulls.size() == 0) {
+    if (!spilled_url.empty()) {
```

Comment:
I think this will fail to restore the object if there is an in-flight pull that later fails, but there is a spilled location that we could've tried here. It'd be good to add a unit test for this scenario.

Reply:
By the way, in #13006, I updated this to store the spilled location in the PullRequest along with the client locations.

Reply:
I think what you're describing was already caught by other unit tests. 

If I understand correctly, the `resend` I just added should deal with it though?

Reply:
Sorry if I missed it. Can you point me to the test that covers this scenario so I can review it?

Reply:
Well after #13006 was merged, this PR was failing `TestManyUpdates`, `TestDedupe`, and `TestRetryTimer`.

I think the scenario you're talking about is 

1. Pull request begins, with the object available on remote node A
2. Pull request is sent to node A
3. node A spills the object then evicts it 
4. Retry timer fires (we should invoke the object spilling code path here).

I think from the pull manager perspective, this is the same as any other change location changes though.